numpyro/hmc_util.py

```diff
@@ -268,3 +271,321 @@ def update_fn(t, accept_prob, z_flat, state):
+
+    _, turning = lax.while_loop(lambda it: (it[0] >= idx_min) & ~it[1],
+                                _body_fn,
+                                (idx_max, False))
+    return turning
+
+
+def _iterative_build_subtree(depth, vv_update, kinetic_fn, z, r, z_grad,
+                             inverse_mass_matrix, step_size, going_right, rng,
+                             energy_current, max_delta_energy, max_tree_depth):
+    max_num_proposals = 2 ** depth
+
+    def _cond_fn(state):
+        tree, turning, _, _, _ = state
+        return (tree.num_proposals < max_num_proposals) & ~turning & ~tree.diverging
+
+    def _body_fn(state):
+        current_tree, _, r_ckpts, r_sum_ckpts, rng = state
+        rng, transition_rng = random.split(rng)
+        z, r, z_grad = _get_leaf(current_tree, going_right)
+        new_leaf = _build_basetree(vv_update, kinetic_fn, z, r, z_grad, step_size, going_right,
+                                   energy_current, max_delta_energy)
+        biased_transition = False
+        new_tree = _combine_tree(current_tree, new_leaf, inverse_mass_matrix, going_right,
+                                 transition_rng, biased_transition, iterative_build=True)
+
+        leaf_idx = current_tree.num_proposals
+        ckpt_idx_min, ckpt_idx_max = _leaf_idx_to_ckpt_idxs(leaf_idx)
+        r, _ = ravel_pytree(new_leaf.r_right)
+        r_sum, _ = ravel_pytree(new_tree.r_sum)
+        # we update checkpoints when leaf_idx is even
+        r_ckpts, r_sum_ckpts = lax.cond(leaf_idx % 2 == 0,
+                                        (r_ckpts, r_sum_ckpts),
+                                        lambda x: (index_update(x[0], ckpt_idx_max, r),
+                                                   index_update(x[1], ckpt_idx_max, r_sum)),
+                                        (r_ckpts, r_sum_ckpts),
+                                        lambda x: x)
+
+        turning = _is_iterative_turning(inverse_mass_matrix, r, r_sum, r_ckpts, r_sum_ckpts,
+                                        ckpt_idx_min, ckpt_idx_max)
+        return new_tree, turning, r_ckpts, r_sum_ckpts, rng
+
+    basetree = _build_basetree(vv_update, kinetic_fn, z, r, z_grad, step_size, going_right,
+                               energy_current, max_delta_energy)
+    r_init, _ = ravel_pytree(basetree.r_left)
+    # TODO: we can create these checkpoints at build_tree method
+    # and reuse it; but let's do this optimization later
+    r_checkpoints = np.zeros((max_tree_depth, inverse_mass_matrix.shape[-1]),
+                             dtype=inverse_mass_matrix.dtype)
+    r_checkpoints = index_update(r_checkpoints, 0, r_init)
+    r_sum_checkpoints = np.zeros((max_tree_depth, inverse_mass_matrix.shape[-1]),
+                                 dtype=inverse_mass_matrix.dtype)
+    r_sum_checkpoints = index_update(r_sum_checkpoints, 0, r_init)
+
+    tree, turning, _, _, _ = lax.while_loop(
```

Comment:
Would `turning` always be true, since we are checking for this in `_cond_fn`? Does this imply that subtree might not be balanced?

Reply:
It does however seem to me that `_double_tree` could return a non-balanced tree when the new tree begins to turn, for instance. Can you confirm if that could happen?

Reply:
Yes, that's how NUTS works: stop doubling (of subtrees) when subtrees are turning. You can take a look at Figure 2 of NUTS paper to see an example for such unbalanced tree.

Reply:
Sorry, I didn't mean that this is a problem per se, but IIRC the tree that is finally returned is always balanced. At what point do we snip off the turning sub tree and return the non turning segment?

Reply:
Okay clarified offline - this doesn't make any difference, since the candidate set will be balanced as per this [line](https://github.com/pyro-ppl/numpyro/pull/54/files#diff-e33a6b3cbaf1b949c73264e0984d9a76R325).